src/db/serde.ts

```diff
@@ -8,25 +16,51 @@ export interface Hook<U, V> {
 const DESERIALIZE_HOOK: Hook<any, any> = {};
 const SERIALIZE_HOOK: Hook<any, any> = {};
 
-function applyHook<T extends Table, U, V>(
-  hook: Hook<U, T>,
-  table: Table,
-  values: InsertableForTable<T> | InsertableForTable<T>[]
-): InsertableForTable<T> | InsertableForTable<T>[] {
-  return Array.isArray(values)
-    ? values.map((v) => applyHookSingle(hook, table, v))
-    : applyHookSingle(hook, table, values);
+type InsertableOrSelectableForTable<T extends Table> =
+  | InsertableForTable<T>
+  | SelectableForTable<T>;
+type InsertableOrSelectableForTableArray<T extends Table> =
+  | InsertableForTable<T>[]
+  | SelectableForTable<T>[];
+
+function applyHook<
+  T extends Table,
+  V extends
+    | InsertableOrSelectableForTable<T>
+    | InsertableOrSelectableForTableArray<T>,
+  U,
+  W
+>(hook: Hook<U, W>, table: Table, values: V, lateral?: FullLateralOption): V {
+  return (
+    Array.isArray(values)
+      ? values.map<V>((v) => applyHookSingle(hook, table, v, lateral))
+      : applyHookSingle(hook, table, values, lateral)
+  ) as V;
```

Comment:
this cast seems weird to me. I guess I need to spend a little more time with applyHookSingle below as well, but let's see...

values is type V (at least according to the typechecker, which may line up with the underlying type). since Array.isArray is true here, that means there is also a type T such that `type V = T[]`. so you do the map, asserting that the map returns a type V, meaning that the type of this `values.map` should be `V[]`...I will spend more time with it but it feels off to me? (thus the cast I think)

Reply:
Isn't it that `V = A | A[]` and `A[] -> A[]` or `A -> A` so `V -> V`?